commit 550dcd53bb6db4f6b9d3b941f101332ac17a0168
Author: Sage Weil <sweil@redhat.com>
Date:   Mon Sep 24 13:03:05 2018 -0500

    mgr/DaemonServer: prevent merge if either pg is remapped|upmap
    
    Remapping means they could be on different OSDs.
    
    Fixes: http://tracker.ceph.com/issues/36166
    Signed-off-by: Sage Weil <sage@redhat.com>

diff --git a/src/mgr/DaemonServer.cc b/src/mgr/DaemonServer.cc
index c643cd8aaf2..374fd4e1e63 100644
--- a/src/mgr/DaemonServer.cc
+++ b/src/mgr/DaemonServer.cc
@@ -2282,6 +2282,23 @@ void DaemonServer::adjust_pgs()
 		pg_t merge_source(p.get_pg_num() - 1, i.first);
 		pg_t merge_target = merge_source.get_parent();
 		bool ok = true;
+
+		if (osdmap.have_pg_upmaps(merge_target)) {
+		  dout(10) << "pool " << i.first
+			   << " pg_num_target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - merge target " << merge_target
+			   << " has upmap" << dendl;
+		  ok = false;
+		} else if (osdmap.have_pg_upmaps(merge_source)) {
+		  dout(10) << "pool " << i.first
+			   << " pg_num_target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - merge source " << merge_source
+			   << " has upmap" << dendl;
+		  ok = false;
+		}
+
 		auto q = pg_map.pg_stat.find(merge_source);
 		if (p.get_pg_num() != p.get_pg_num_pending()) {
 		  dout(10) << "pool " << i.first
@@ -2315,7 +2332,15 @@ void DaemonServer::adjust_pgs()
 			   << " not clean (" << pg_state_string(q->second.state)
 			   << ")" << dendl;
 		  ok = false;
+		} else if (q->second.state & PG_STATE_REMAPPED) {
+		  dout(10) << "pool " << i.first
+			   << " pg_num_target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - merge source " << merge_source
+			   << " remapped" << dendl;
+		  ok = false;
 		}
+
 		q = pg_map.pg_stat.find(merge_target);
 		if (q == pg_map.pg_stat.end()) {
 		  dout(10) << "pool " << i.first
@@ -2334,6 +2359,13 @@ void DaemonServer::adjust_pgs()
 			   << " not clean (" << pg_state_string(q->second.state)
 			   << ")" << dendl;
 		  ok = false;
+		} else if (q->second.state & PG_STATE_REMAPPED) {
+		  dout(10) << "pool " << i.first
+			   << " pg_num_target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - merge target " << merge_target
+			   << " remapped" << dendl;
+		  ok = false;
 		}
 		if (ok) {
 		  unsigned target = p.get_pg_num() - 1;
